### PR TITLE
Refactor MOTPE weighting

### PR DESCRIPTION
**Pull request by nabenabe0928**
## Motivation

This PR was separated from:
- https://github.com/optuna/optuna/pull/5813

Basically, this PR aims to make the TPE implementation shorter so that reusability increases.

## Description of the changes

- Refactor the MOTPE weighting algorithm

## NOTE

From the monotonicity of the hypervolume (submodularity), `hv` is always larger than the leave-one-out hypervolume, i.e. `compute_hypervolume(lvals[indices_mat[i]], reference_point)`, meaning that each element of `contributions` becomes `inf` only if `hv` is `inf`.

Since we also define the line `contributions[np.isnan(contributions)] = np.inf`, which takes care of `inf - inf = inf` in the hypervolume contribution calculation, if any element in `contributions` takes `inf`, all the elements in `contributions` become `inf`.

That is why we can separate this exception handling as the processing only for `np.isinf(hv) == True`.

### CODE REVIEW

**Comment by nabenabe0928** [comment]
@y0z @kAIto47802 
Could you review this this PR when you have time?

File: optuna/samplers/_tpe/sampler.py [hunk 1]
```diff
@@ -742,52 +750,30 @@ def _calculate_weights_below_for_multi_objective(
...
-                hv - compute_hypervolume(lvals[indices_mat[i]], reference_point)
-                for i in range(n_below)
-            ]
-        )
-        contributions[np.isnan(contributions)] = np.inf
-        max_contribution = np.maximum(np.max(contributions), EPS)
-        if not np.isfinite(max_contribution):
-            weights_below = np.ones_like(contributions, dtype=float)
-            # TODO(nabenabe0928): Make the weights for non Pareto solutions to zero.
-            weights_below[np.isfinite(contributions)] = EPS
```
Comment (nabenabe0928) [inline optuna/samplers/_tpe/sampler.py:786]: As mentioned here,
- https://github.com/optuna/optuna/pull/5871#issue-2749413723

if `hv` is `inf`, all the elements in `contributions` also become `inf`.

Let's define this array as `hvs`:

```python
hvs = np.array([compute_hypervolume(lvals[indices_mat[i]], reference_point) for i in range(n_below)])
```

If `hvs[i]` is `inf`, `hv - hvs[i]` becomes `nan`, leading to `contributions[i] = np.inf` due to `contributions[np.isnan(contributions)] = np.inf`.

If `hvs[i]` is finite, `hv - hvs[i]` anyways becomes `inf`, so `contributions[i] = np.inf`.

It simply means that if `hv` is `inf`, all the elements in `contributions` are `inf`.

That is why `np.isfinite(contributions)` is all False, so all the elements in `weights_below` are 1.

File: optuna/samplers/_tpe/sampler.py [hunk 2]
```diff
@@ -742,52 +750,30 @@ def _calculate_weights_below_for_multi_objective(
...
+    lvals = np.asarray([t.values for t in below_trials])[is_feasible]
+    lvals *= np.array([-1.0 if d == StudyDirection.MAXIMIZE else 1.0 for d in study.directions])
+    ref_point = _get_reference_point(lvals)
+    on_front = _is_pareto_front(lvals, assume_unique_lexsorted=False)
+    pareto_sols = lvals[on_front]
+    hv = compute_hypervolume(pareto_sols, ref_point, assume_pareto=True)
+    if np.isinf(hv):
+        # TODO(nabenabe): Assign EPS to non-Pareto solutions, and
+        # solutions with finite contrib if hv is inf. Ref: PR#5813.
+        return weights_below
```
Comment (nabenabe0928) [inline optuna/samplers/_tpe/sampler.py:771]: Maybe we should, first of all, replace $\pm \infty$ with a big value like here:

- https://github.com/optuna/optuna/pull/5798/files#diff-152bf351e020a2e950295910a479597f1026e376322ef4c934e4f04ad6f69fafR119-R124

**Review by kAIto47802 — APPROVED** [review]
Thank you for your PR. Nice implementation! LGTM!

**Comment by nabenabe0928** [comment]
@y0z
If you have time, could you take another look at this PR?

**Review by y0z — APPROVED** [review]
LGTM

**Pull request merged**